Output average eval from threads suggesting chosen move

xoto10 opened the pull request:
If chosen move was suggested by 3 threads with evals 45, 60 and 60, current code randomly displays 45 or 60 for the eval. Display the average (55) instead.
This can make a large difference in some cases if an outlier eval is displayed instead of the value found by most other threads. 

STC 5+0.05 th 4 :
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 142824 W: 30368 L: 30481 D: 81975
http://tests.stockfishchess.org/tests/view/5cb611e60ebc5925cf01a4d9

Bench 3402947

mcostalba commented:
Current code selects the best thread, according to the selction algorithm and then prints its PV and its score. You are suggesting to replace this by still choosing a best thread and prints its PV, but the score instead if from a mix of threads that have the same best move.

May be selction alghorithm could be improved, but the current design is very simple, pick the best thread and print its score/PV. I don't see your patch is improving in anyway apart from messing further up a part of code that, due to selection alghortim, is already quite complicated. I also prefer to have the PV and the score in sync.

Considering that your proposed advantage is purely speculative, I am closing this one.

xoto10 commented:
Thanks, your comments have made me look at the PV more as well as the eval.

The current algorithm works in 2 stages, first, vote for the best move, and then choose one of the threads using that move. The move choice is obviously working well, my issue is with the choosing of a thread - currently it just takes the first thread it finds with the right move.

I have updated the code with debugging info. These are some illustrations of what I see as problems with the current code:
```
<< info depth 14 seldepth 23 multipv 1 score cp 135 nodes 168729 nps 5272781 tbhits 0 time 32 pv f2d2 f5f4 b4b5 g4f5 c2b3 g6g5 b5a6 f5d7 c5c6 d7g4 b3a4 h8a8 a4b5 f4f3
<< info string       first move f2d2
<< info string th 0                  cp 135
<< info string th 1                  cp 172
<< info string th 2                  cp 172
<< bestmove f2d2 ponder f5f4

<< info depth 15 seldepth 26 multipv 1 score cp -151 nodes 665918 nps 5643372 tbhits 0 time 118 pv c6e5 c4f1 g7f6 b3b4 f5f4 b4b5 e5d3 a5a6 b7a6 b5a6 e7a7 a1a4 f6f5 f1e2 f5e5
<< info string       first move c6e5
<< info string th 0                  cp -151
<< info string th 1 better move e7c7 cp -148
<< info string th 2                  cp -138
<< info depth 15 seldepth 24 multipv 1 score cp -148 nodes 666058 nps 5644559 tbhits 0 time 118 pv e7c7 c4f1 g7f6 b3b4 c6e5 b4b5 f6e6 a1a3 e6d6 b5b6 c7e7 a3b3 d6c6 f1b5 c6c5 b5a4 e5c6 a5a6 b7a6 b6b7 e4e3 b7b8q c6b8 b3b8
<< bestmove e7c7 ponder c4f1
```
The score and pv displayed comes from the 1st thread found with the correct move, even though a later thread has a better score, and will contribute more to the voting algorithm. These differences could be large, presumably right up to piece capture or game winning/losing size in extreme cases.


I have commented out the eval averaging line in my code and added an if statement to update bestThread if a higher score is found. Some example debug with this new code:
```
<< info depth 13 seldepth 20 multipv 1 score cp 23 nodes 410791 nps 4234958 tbhits 0 time 97 pv h2h4 b8c6 e1g1 d7g4 b2b4 c6b4 d2b4 e7b4 f3e5 b4f4 e5g4 h5g4
<< info string       first move h2h4
<< info string th 0                  cp 23
<< info string th 1                  cp 38 new best thread
<< info string th 2                  cp 50 new best thread
<< info depth 13 seldepth 19 multipv 1 score cp 50 nodes 411222 nps 4196142 tbhits 0 time 98 pv h2h4 b8c6 e1g1 d7g4 c2c3 e7d7 a2a4 g4f3 e2f3 d7d3 d2g5 d3d1 f1d1 f8e7 b2b4 e8g8 g3h5 f6h5
<< bestmove h2h4 ponder b8c6

<< info depth 12 seldepth 19 multipv 1 score cp 57 nodes 76467 nps 3823350 tbhits 0 time 20 pv d8c7 b2b4 g4h2 f3h2 c7g3 h2f1 g3d3 g2e4 d3h3 d1f3 h3f3 e4f3 d4d3 f3c6 b7c6 c2d4
<< info string       first move d8c7
<< info string th 0                  cp 57
<< info string th 1                  cp 81 new best thread
<< info string th 2                  cp 97 new best thread
<< info depth 12 seldepth 24 multipv 1 score cp 97 nodes 77089 nps 3854450 tbhits 0 time 20 pv d8c7 b2b4 g4h2 f3h2 c7g3 h2f1 g3d3 g2e4 d3h3 d1f3 h3f3 e4f3 d4d3 f3c6 b7c6 c2d4
<< bestmove d8c7 ponder b2b4
```
Here the thread with the best score is chosen, so a higher score and different pv are output. We're already looping through all the threads here, I think the code should try a bit harder to pick the best thread, not just the first one it gets to.

My personal preference is still to display an averaged score. The voting algorithm means we're not really choosing a single pv, we're choosing a first move with a range of attractive pv's. We have to choose one to display last for guis to use, often there is no "correct" one to choose, we just have to pick one. 
But for the score, we have a group of estimated scores. If there are 20 threads all saying f2d2 as in the first example above, some saying cp 135 and some saying 172, it seems wrong to display the optimistic 172 if 19 out of 20 think the eval is 135. On the other hand, if 19 say it is 172, then we should display 172 or something close to it. Averaging is an easy way to combine the scores from several possible pvs into something meaningful to the user, even if we can't do something similar with the pvs.

xoto10 commented:
Updated code with debug and averaging statement commented out can be seen [here](https://github.com/official-stockfish/Stockfish/compare/master...xoto10:best1) for anyone wanting to play with this.
